incubator/hnc/internal/reconcilers/object.go

```diff
@@ -359,9 +360,26 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
 	if srcInst == nil {
 		return actionRemove, nil
 	}
-
 	// If an object doesn't exist, assume it's been deleted or not yet created.
 	exists := inst.GetCreationTimestamp() != v1.Time{}
+	selector := r.getSelector(ctx, log, srcInst)
+
+	nsLabels := labels.Set(r.Forest.Get(inst.GetNamespace()).GetLabels())
+	if selector != nil && !selector.Matches(nsLabels) {
+		if exists {
+			// The object already exists but it shouldn't be here according to the updated selector, so we remove it
+			r.deleteObject(ctx, log, inst)
+			return actionRemove, nil
```

Comment:
Returning `actionRemove` tells the _caller_ that it needs to delete the object, so don't call `r.deleteObject` here. We never make any commands to the apiserver while we hold the global mutex since network calls are slow (seconds) whereas we want to be able to lock/unlock the mutex hundreds of times a second.

Also, you don't have any error handling here, but `r.deleteObject()` return an error and we should never just ignore errors (without a very good reason why).